File: src/shapes/omp.ts
import { createReadStream } from "node:fs";
import { createInterface } from "node:readline";
import { basename } from "node:path";
import { asObj, str, truncate, uidOf, type ParsedEvent, type ParsedFile, type Parser } from "../types.js";

/**
 * omp (GLM/Claude-family coding agent) transcripts —
 * ~/.omp/agent/sessions/<encoded-cwd>/<timestamp>_<id>.jsonl
 *
 * Discovered 2026-09-18 while teaching a live omp pane to use relic. The prior source
 * survey in sources.ts called this "SQLITE, not JSONL" and excluded it — wrong: the
 * per-session transcripts sitting right next to history.db ARE plain JSONL, one event
 * per line. That survey error is why omp calling `relic_now` always resolved to
 * whichever OTHER agent's Claude Code session was newest in the same directory — omp
 * had no source of its own to be found in.
 *
 * Shape differs from Claude Code in the details but carries the same information:
 *   role 'user'       -> user prose, content is always an array of blocks (never bare string)
 *   role 'assistant'  -> content blocks: 'text' | 'thinking' | 'toolCall'
 *   role 'toolResult' -> its own envelope role (not a tool_result BLOCK inside 'user',
 *                        the way Claude nests it) — so no blockRole() reclassification
 *                        is needed here; the envelope already says what it is.
 * toolName/toolCallId live on the message envelope for a toolResult, and name/arguments
 * live directly on a 'toolCall' block — neither matches Claude's tool_use/tool_result
 * field layout, so this gets its own flatten function rather than reusing
 * flattenContent/blockRole, which are typed to Claude's exact block vocabulary.
 */

const INDEXED = new Set(["user", "assistant", "toolResult"]);

function flattenOmp(content: unknown): { text: string; role: string | null } {
  if (!Array.isArray(content)) return { text: "", role: null };
  const kinds = new Set<string>();
  const out: string[] = [];
  for (const raw of content) {
    const item = asObj(raw);
    if (!item) continue;
    const t = str(item.type);
    if (t) kinds.add(t);
    if (t === "text") {
      const txt = str(item.text);
      if (txt) out.push(txt);
    } else if (t === "thinking") {
      const th = str(item.thinking);
      if (th) out.push(th);
    } else if (t === "toolCall") {
      out.push(`[tool_use ${str(item.name) ?? "?"}] ${truncate(JSON.stringify(item.arguments ?? {}), 2000)}`);
    }
  }
  // Same "label by what the block actually is" rule as Claude — see blockRole()'s
  // comment in types.ts. A content array holding ONLY toolCall/thinking blocks is
  // tool traffic or reasoning, not conversation, whichever envelope role carried it.
  let role: string | null = null;
  if (kinds.has("toolCall") && !kinds.has("text")) role = "tool_use";
  else if (kinds.has("thinking") && kinds.size === 1) role = "thinking";
  return { text: out.join("\n"), role };
}

export const parseOmp: Parser = async (filePath) => {
  const fileKey = basename(filePath);
  const events: ParsedEvent[] = [];
  const typeCounts: Record<string, number> = {};
  let lines = 0, badLines = 0, seq = 0;
  let cwd: string | null = null, model: string | null = null;
  // omp's own id field is the session identity — the filename ALSO embeds it after
  // the timestamp prefix, but the "session" record is the source of truth.
  let sessionUuid = fileKey.replace(/\.jsonl$/, "");
  let startedAt: string | null = null, endedAt: string | null = null;
  let description: string | null = null, title: string | null = null;

  const rl = createInterface({ input: createReadStream(filePath, "utf8") });
  for await (const line of rl) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    lines++; seq++;

    let parsed: unknown;
    try { parsed = JSON.parse(trimmed); } catch { badLines++; continue; }
    const rec = asObj(parsed);
    if (!rec) { badLines++; continue; }

    const type = str(rec.type) ?? "unknown";
    typeCounts[type] = (typeCounts[type] ?? 0) + 1;

    if (type === "session") {
      if (!cwd) cwd = str(rec.cwd);
      const id = str(rec.id);
      if (id) sessionUuid = id;
    }
    if (type === "model_change" && !model) model = str(rec.model);
    // Overwrite, never keep-first — same reasoning as Claude's ai-title: the title
    // is refined as the session goes on, so the LAST one names what it became.
    if ((type === "title" || type === "title_change") && !title) {
      const t = str(rec.title);
      if (t) title = t;
    }

    const ts = str(rec.timestamp);
    if (ts) { if (!startedAt) startedAt = ts; endedAt = ts; }

    if (!INDEXED.has(type === "message" ? str(asObj(rec.message)?.role) ?? "" : "")) continue;

    const msg = asObj(rec.message);
    const envelopeRole = str(msg?.role);
    if (!envelopeRole) continue;

    const { text, role: blockRole } = flattenOmp(msg?.content);
    if (!text) continue;

    const role = blockRole ?? (envelopeRole === "toolResult" ? "tool_result" : envelopeRole);

    if (!description && role === "user") description = truncate(text, 200);
    events.push({
      uid: uidOf("omp", fileKey, seq),
      seq, role, ts, text: truncate(text),
    });
  }

  return { sessionUuid, cwd, model, events, lines, badLines, typeCounts, startedAt, endedAt,
           description, title, gitBranch: null };
};

File: python/relicpy/shapes/omp.py
"""omp (Oh My Pi) — ~/.omp/agent/sessions/<encoded-cwd>/<timestamp>_<id>.jsonl."""

from __future__ import annotations

import json
import os
import re

from ..models import ParsedEvent, ParsedFile
from ..types import _compact_json, as_obj, s, truncate, uid_of

INDEXED = {"user", "assistant", "toolResult"}


def _flatten_omp(content) -> tuple[str, str | None]:
    """Same "label by what the block actually is" rule as Claude.

    A content array holding ONLY toolCall/thinking blocks is tool traffic or reasoning,
    not conversation, whichever envelope role carried it.
    """
    if not isinstance(content, list):
        return "", None
    kinds, out = set(), []
    for raw in content:
        item = as_obj(raw)
        if not item:
            continue
        t = s(item.get("type"))
        if t:
            kinds.add(t)
        if t == "text":
            txt = s(item.get("text"))
            if txt:
                out.append(txt)
        elif t == "thinking":
            th = s(item.get("thinking"))
            if th:
                out.append(th)
        elif t == "toolCall":
            name = s(item.get("name")) or "?"
            out.append(f"[tool_use {name}] {truncate(_compact_json(item.get('arguments') or {}), 2000)}")
    role = None
    if "toolCall" in kinds and "text" not in kinds:
        role = "tool_use"
    elif kinds == {"thinking"}:
        role = "thinking"
    return "\n".join(out), role


def parse(file_path: str) -> ParsedFile:
    file_key = os.path.basename(file_path)
    events: list[ParsedEvent] = []
    lines = bad_lines = seq = 0
    cwd = model = title = None
    session_uuid = re.sub(r"\.jsonl$", "", file_key)
    started_at = ended_at = description = None

    with open(file_path, "r", encoding="utf-8", errors="replace") as fh:
        for line in fh:
            trimmed = line.strip()
            if not trimmed:
                continue
            lines += 1
            seq += 1
            try:
                rec = as_obj(json.loads(trimmed))
            except Exception:
                bad_lines += 1
                continue
            if rec is None:
                bad_lines += 1
                continue

            typ = s(rec.get("type")) or "unknown"
            if typ == "session":
                if not cwd:
                    cwd = s(rec.get("cwd"))
                sid = s(rec.get("id"))
                if sid:
                    session_uuid = sid
            if typ == "model_change" and not model:
                model = s(rec.get("model"))
            if typ in ("title", "title_change") and not title:
                t = s(rec.get("title"))
                if t:
                    title = t

            ts = s(rec.get("timestamp"))
            if ts:
                if not started_at:
                    started_at = ts
                ended_at = ts

            msg = as_obj(rec.get("message"))
            envelope_role = (s(msg.get("role")) if msg else None)
            gate = envelope_role or "" if typ == "message" else ""
            if gate not in INDEXED:
                continue
            if not envelope_role:
                continue

            text, brole = _flatten_omp(msg.get("content") if msg else None)
            if not text:
                continue
            role = brole or ("tool_result" if envelope_role == "toolResult" else envelope_role)

            if not description and role == "user":
                description = truncate(text, 200)
            events.append(ParsedEvent(uid=uid_of("omp", file_key, seq), seq=seq,
                                      role=role, ts=ts, text=truncate(text)))

    return ParsedFile(session_uuid=session_uuid, cwd=cwd, model=model, events=events,
                      lines=lines, bad_lines=bad_lines, started_at=started_at,
                      ended_at=ended_at, description=description, title=title, git_branch=None)
